It is caused by using removed API in Gradle 9:

```
 error: constructor DefaultProjectDependency in class DefaultProjectDependency cannot be applied to given types;
                new DefaultProjectDependency((ProjectInternal) project, true, taskDependencyFactory));
                ^
  required: ProjectState
  found:    ProjectInternal,boolean,TaskDependencyFactory
  reason: actual and formal argument lists differ in length
```
```
error: cannot find symbol
        return resolvedArtifacts.size() < configuration.getFiles().size()
                                                       ^
  symbol:   method getFiles()
  location: variable configuration of type ResolvedConfiguration
2 errors
2 warnings
```

--- a/.github/workflows/ci-actions-incremental.yml
+++ b/.github/workflows/ci-actions-incremental.yml
@@ -773,7 +773,7 @@ jobs:
         env:
           CAPTURE_BUILD_SCAN: true
         # Important: keep -pl ... in sync with "Calculate run flags"!
-        run: ./mvnw $COMMON_MAVEN_ARGS $COMMON_TEST_MAVEN_ARGS $PTS_MAVEN_ARGS $JVM_TEST_MAVEN_ARGS clean install -pl integration-tests/gradle
+        run: ./mvnw $COMMON_MAVEN_ARGS $COMMON_TEST_MAVEN_ARGS $PTS_MAVEN_ARGS $JVM_TEST_MAVEN_ARGS clean install -pl integration-tests/gradle -Pgradle-8
       - name: Prepare build reports archive
         if: always()
         run: |

--- a/build-parent/pom.xml
+++ b/build-parent/pom.xml
@@ -62,7 +62,7 @@
         <!-- These 2 properties are used by CreateProjectMojo to add the Maven Wrapper -->
         <proposed-maven-version>3.9.9</proposed-maven-version>
         <maven-wrapper.version>3.3.2</maven-wrapper.version>
-        <gradle-wrapper.version>8.14</gradle-wrapper.version>
+        <gradle-wrapper.version>9.0.0</gradle-wrapper.version>
         <quarkus-gradle-plugin.version>${project.version}</quarkus-gradle-plugin.version>
         <quarkus-maven-plugin.version>${project.version}</quarkus-maven-plugin.version>
         <maven-plugin-plugin.version>3.8.1</maven-plugin-plugin.version>

--- a/devtools/gradle/gradle/wrapper/gradle-wrapper.properties
+++ b/devtools/gradle/gradle/wrapper/gradle-wrapper.properties
@@ -1,8 +1,8 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 # https://gradle.org/release-checksums/
-distributionSha256Sum=efe9a3d147d948d7528a9887fa35abcf24ca1a43ad06439996490f77569b02d1
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.14-all.zip
+distributionSha256Sum=f759b8dd5204e2e3fa4ca3e73f452f087153cf81bac9561eeb854229cc2c5365
+distributionUrl=https\://services.gradle.org/distributions/gradle-9.0.0-all.zip
 networkTimeout=10000
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists

--- a/independent-projects/bootstrap/pom.xml
+++ b/independent-projects/bootstrap/pom.xml
@@ -71,7 +71,7 @@
         <plexus-utils.version>3.5.1</plexus-utils.version>
         <smallrye-common.version>2.13.8</smallrye-common.version>
         <smallrye-beanbag.version>1.5.3</smallrye-beanbag.version>
-        <gradle-tooling.version>8.14</gradle-tooling.version>
+        <gradle-tooling.version>9.0.0</gradle-tooling.version>
         <quarkus-fs-util.version>1.1.0</quarkus-fs-util.version>
         <crac.version>1.5.0</crac.version>
     </properties>

--- a/independent-projects/tools/devtools-testing/src/main/resources/fake-catalog.json
+++ b/independent-projects/tools/devtools-testing/src/main/resources/fake-catalog.json
@@ -453,7 +453,7 @@
         "recommended-java-version": "17",
         "proposed-maven-version": "3.9.9",
         "maven-wrapper-version": "3.3.2",
-        "gradle-wrapper-version": "8.14"
+        "gradle-wrapper-version": "9.0.0"
       }
     },
     "codestarts-artifacts": [

--- a/independent-projects/tools/pom.xml
+++ b/independent-projects/tools/pom.xml
@@ -41,7 +41,7 @@
         <!-- When updating gradle-wrapper.version, make sure to also update files in
              independent-projects/tools/base-codestarts/src/main/resources/codestarts
              /quarkus/tooling/gradle-wrapper/base/gradle -->
-        <gradle-wrapper.version>8.14</gradle-wrapper.version>
+        <gradle-wrapper.version>9.0.0</gradle-wrapper.version>
 
         <!-- These properties are needed in order for them to be resolvable by the generated projects -->
         <compiler-plugin.version>3.14.0</compiler-plugin.version>

--- a/independent-projects/tools/registry-client/src/test/resources/catalog-config/quarkus-bom-quarkus-platform-descriptor-999-SNAPSHOT-999-SNAPSHOT.json
+++ b/independent-projects/tools/registry-client/src/test/resources/catalog-config/quarkus-bom-quarkus-platform-descriptor-999-SNAPSHOT-999-SNAPSHOT.json
@@ -3048,7 +3048,7 @@
         "supported-maven-versions" : "[3.6.2,)",
         "proposed-maven-version" : "3.8.1",
         "maven-wrapper-version" : "3.3.2",
-        "gradle-wrapper-version" : "7.2"
+        "gradle-wrapper-version" : "9.0.0"
       }
     },
     "codestarts-artifacts" : [ "io.quarkus:quarkus-project-core-extension-codestarts::jar:999-SNAPSHOT" ]

--- a/independent-projects/tools/registry-client/src/test/resources/catalog-config/quarkus-universe-bom-quarkus-platform-descriptor-2.0.3.Final-2.0.3.Final.json
+++ b/independent-projects/tools/registry-client/src/test/resources/catalog-config/quarkus-universe-bom-quarkus-platform-descriptor-2.0.3.Final-2.0.3.Final.json
@@ -7297,7 +7297,7 @@
         "supported-maven-versions" : "[3.6.2,)",
         "proposed-maven-version" : "3.8.1",
         "maven-wrapper-version" : "3.3.2",
-        "gradle-wrapper-version" : "6.9"
+        "gradle-wrapper-version" : "9.0.0"
       }
     },
     "codestarts-artifacts" : [ "io.quarkus:quarkus-project-core-extension-codestarts::jar:2.0.3.Final" ],

--- a/integration-tests/gradle/gradle/wrapper/gradle-wrapper.properties
+++ b/integration-tests/gradle/gradle/wrapper/gradle-wrapper.properties
@@ -1,8 +1,8 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 # https://gradle.org/release-checksums/
-distributionSha256Sum=efe9a3d147d948d7528a9887fa35abcf24ca1a43ad06439996490f77569b02d1
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.14-all.zip
+distributionSha256Sum=f759b8dd5204e2e3fa4ca3e73f452f087153cf81bac9561eeb854229cc2c5365
+distributionUrl=https\://services.gradle.org/distributions/gradle-9.0.0-all.zip
 networkTimeout=10000
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists

--- a/integration-tests/gradle/pom.xml
+++ b/integration-tests/gradle/pom.xml
@@ -545,6 +545,37 @@
             </properties>
         </profile>
         <profile>
+            <id>gradle-8</id>
+            <!-- Adds an execution to the surefire plugin that runs selected tests with Gradle 8 -->
+            <build>
+                <pluginManagement>
+                    <plugins>
+                        <plugin>
+                            <groupId>org.apache.maven.plugins</groupId>
+                            <artifactId>maven-surefire-plugin</artifactId>
+                            <executions>
+                                <execution>
+                                    <id>gradle-wrapper-8.14</id>
+                                    <phase>test</phase>
+                                    <goals>
+                                        <goal>test</goal>
+                                    </goals>
+                                    <configuration>
+                                        <systemPropertyVariables>
+                                            <quarkus-test-gradle-wrapper-version>8.14</quarkus-test-gradle-wrapper-version>
+                                        </systemPropertyVariables>
+                                        <includes>
+                                            <include>BasicMultiModuleQuarkusBuildTest</include>
+                                        </includes>
+                                    </configuration>
+                                </execution>
+                            </executions>
+                        </plugin>
+                    </plugins>
+                </pluginManagement>
+            </build>
+        </profile>
+        <profile>
             <id>native-image</id>
             <!-- Since there is no quarkus-maven-plugin build goal in this pom,
                  this profile will not directly yield a native image like in the other modules.

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/CompileOnlyDependencyFlagsTest.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/CompileOnlyDependencyFlagsTest.java
@@ -20,11 +20,11 @@ import io.quarkus.maven.dependency.ArtifactCoords;
 import io.quarkus.maven.dependency.DependencyFlags;
 import io.quarkus.runtime.LaunchMode;
 
-public class CompileOnlyDependencyFlagsTest {
+public class CompileOnlyDependencyFlagsTest extends QuarkusGradleTestBase {
 
     @Test
     public void compileOnlyFlags() throws Exception {
-        var projectDir = QuarkusGradleTestBase.getProjectDir("compile-only-dependency-flags");
+        var projectDir = getProjectDir("compile-only-dependency-flags");
 
         final String componly = ArtifactCoords.jar("org.acme", "componly", "1.0.0-SNAPSHOT").toCompactCoords();
         final String common = ArtifactCoords.jar("org.acme", "common", "1.0.0-SNAPSHOT").toCompactCoords();

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/QuarkusGradleTestBase.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/QuarkusGradleTestBase.java
@@ -6,27 +6,50 @@ import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.io.UncheckedIOException;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
 import java.util.Properties;
+
+import io.quarkus.bootstrap.util.IoUtils;
 
 public class QuarkusGradleTestBase {
 
-    protected static File getProjectDir(final String projectName)
-            throws URISyntaxException, IOException {
-        final URL projectUrl = Thread.currentThread().getContextClassLoader().getResource(projectName);
+    protected File getProjectDir(final String projectName) {
+        return getProjectDir(projectName, null);
+    }
+
+    protected File getProjectDir(final String baseProjectName, String testSuffix) {
+        final URL projectUrl = Thread.currentThread().getContextClassLoader().getResource(baseProjectName);
         if (projectUrl == null) {
-            throw new IllegalStateException("Failed to locate test project " + projectName);
+            throw new IllegalStateException("Failed to locate test project " + baseProjectName);
         }
-        final File projectDir = new File(projectUrl.toURI());
+        File projectDir;
+        try {
+            projectDir = new File(projectUrl.toURI());
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
         if (!projectDir.isDirectory()) {
             throw new IllegalStateException(projectDir + " is not a directory");
+        }
+
+        if (testSuffix != null && !testSuffix.isBlank()) {
+            var testProjectName = new StringBuilder().append(projectDir.getName());
+            if (testSuffix.charAt(0) != '-') {
+                testProjectName.append("-");
+            }
+            testProjectName.append(testSuffix);
+            Path targetDir = projectDir.getParentFile().toPath().resolve(testProjectName.toString());
+            try {
+                IoUtils.copy(projectDir.toPath(), targetDir);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            projectDir = targetDir.toFile();
         }
 
         final Properties props = new Properties();
@@ -34,6 +57,8 @@ public class QuarkusGradleTestBase {
         if (projectProps.exists()) {
             try (InputStream is = new FileInputStream(projectProps)) {
                 props.load(is);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
             }
         } else {
             props.setProperty("quarkusPlatformGroupId", "io.quarkus");
@@ -45,31 +70,25 @@ public class QuarkusGradleTestBase {
         props.setProperty("quarkusPluginVersion", quarkusVersion);
         try (OutputStream os = new FileOutputStream(projectProps)) {
             props.store(os, "Quarkus Gradle TS");
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
         }
         return projectDir;
     }
 
-    protected static String getQuarkusVersion() throws IOException {
+    protected static String getQuarkusVersion() {
         final Path curDir = Paths.get("").toAbsolutePath().normalize();
         final Path gradlePropsFile = curDir.resolve("gradle.properties");
         Properties props = new Properties();
         try (InputStream is = Files.newInputStream(gradlePropsFile)) {
             props.load(is);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
         }
         final String quarkusVersion = props.getProperty("version");
         if (quarkusVersion == null) {
             throw new IllegalStateException("Failed to locate Quarkus version in " + gradlePropsFile);
         }
         return quarkusVersion;
-    }
-
-    protected List<String> arguments(String... argument) {
-        List<String> arguments = new ArrayList<>();
-        arguments.addAll(Arrays.asList(argument));
-        String mavenRepoLocal = System.getProperty("maven.repo.local", System.getenv("MAVEN_LOCAL_REPO"));
-        if (mavenRepoLocal != null) {
-            arguments.add("-Dmaven.repo.local=" + mavenRepoLocal);
-        }
-        return arguments;
     }
 }

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/QuarkusGradleWrapperTestBase.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/QuarkusGradleWrapperTestBase.java
@@ -6,7 +6,6 @@ import java.io.File;
 import java.io.FileWriter;
 import java.io.IOException;
 import java.lang.management.ManagementFactory;
-import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashMap;
@@ -22,11 +21,38 @@ public class QuarkusGradleWrapperTestBase extends QuarkusGradleTestBase {
     private static final String GRADLE_WRAPPER_WINDOWS = "gradlew.bat";
     private static final String GRADLE_WRAPPER_UNIX = "./gradlew";
     private static final String MAVEN_REPO_LOCAL = "maven.repo.local";
+    private static final String QUARKUS_TEST_GRADLE_WRAPPER_VERSION = "quarkus-test-gradle-wrapper-version";
 
     private Map<String, String> systemProps;
 
     private boolean configurationCacheEnable = true;
     private boolean noWatchFs = true;
+
+    @Override
+    protected File getProjectDir(String projectName) {
+        return getProjectDir(projectName, getTestProjectNameSuffix());
+    }
+
+    /**
+     * Returns a suffix for a test project directory name based on the requested Gradle wrapper version.
+     * If a Gradle wrapper version was not configured, the suffix will be null.
+     * Otherwise, it will be {@code -wrapper-${quarkus-test-gradle-wrapper-version}}.
+     *
+     * @return test project directory suffix or null
+     */
+    private static String getTestProjectNameSuffix() {
+        var wrapperVersion = getRequestedWrapperVersion();
+        return wrapperVersion == null ? null : "-wrapper-" + wrapperVersion;
+    }
+
+    /**
+     * Returns configured Gradle wrapper version for a test or null, in case it wasn't configured.
+     *
+     * @return configured Gradle wrapper version or null
+     */
+    private static String getRequestedWrapperVersion() {
+        return System.getProperty(QUARKUS_TEST_GRADLE_WRAPPER_VERSION);
+    }
 
     protected void setupTestCommand() {
 
@@ -61,9 +87,12 @@ public class QuarkusGradleWrapperTestBase extends QuarkusGradleTestBase {
             throws IOException, InterruptedException {
         boolean isInCiPipeline = "true".equals(System.getenv("CI"));
 
+        // install a custom version of the wrapper, in case it's configured
+        installRequestedWrapper(projectDir);
+
         setupTestCommand();
         List<String> command = new ArrayList<>();
-        command.add(getGradleWrapperCommand());
+        command.add(getGradleWrapperCommand(projectDir));
         addSystemProperties(command);
 
         if (!isInCiPipeline && isDebuggerConnected()) {
@@ -87,7 +116,7 @@ public class QuarkusGradleWrapperTestBase extends QuarkusGradleTestBase {
 
         File logOutput = new File(projectDir, "command-output.log");
 
-        System.out.println("$ " + String.join(" ", command));
+        logCommandLine(command);
         ProcessBuilder pb = new ProcessBuilder()
                 .directory(projectDir)
                 .command(command)
@@ -142,6 +171,34 @@ public class QuarkusGradleWrapperTestBase extends QuarkusGradleTestBase {
         return commandResult;
     }
 
+    private static void logCommandLine(List<String> command) {
+        System.out.println("$ " + String.join(" ", command));
+    }
+
+    private void installRequestedWrapper(File projectDir) {
+        String wrapperVersion = getRequestedWrapperVersion();
+        if (wrapperVersion == null) {
+            // no specific version was configured, the integration-test/gradle one will be used
+            return;
+        }
+        final String defaultWrapper = getGradleWrapperCommand(projectDir);
+
+        final List<String> command = List.of(defaultWrapper, "wrapper", "--gradle-version=" + wrapperVersion);
+        logCommandLine(command);
+
+        final ProcessBuilder pb = new ProcessBuilder()
+                .directory(projectDir)
+                .command(command)
+                .redirectInput(ProcessBuilder.Redirect.INHERIT)
+                // Should prevent "fragmented" output (parts of stdout and stderr interleaved)
+                .redirectErrorStream(true);
+        try {
+            pb.start().waitFor();
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to install Gradle wrapper", e);
+        }
+    }
+
     protected void setSystemProperty(String name, String value) {
         if (systemProps == null) {
             systemProps = new HashMap<>();
@@ -149,8 +206,15 @@ public class QuarkusGradleWrapperTestBase extends QuarkusGradleTestBase {
         systemProps.put(name, value);
     }
 
-    private String getGradleWrapperCommand() {
-        return Paths.get(getGradleWrapperName()).toAbsolutePath().toString();
+    private String getGradleWrapperCommand(File projectDir) {
+        File wrapper = null;
+        if (projectDir != null) {
+            wrapper = new File(projectDir, getGradleWrapperName());
+        }
+        if (wrapper == null || !wrapper.exists()) {
+            wrapper = new File(getGradleWrapperName());
+        }
+        return wrapper.getAbsolutePath();
     }
 
     private String getGradleWrapperName() {

